File: communication.py
# Copyright (c) 2021 MIT
# 
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

import torch
import torch.distributed as dist
from logger import Logger

class CommunicationBackend:
    def __init__(self, rank: int, world_size: int, master_addr: str, master_port: int, backend: str, device='cuda:0'):
        self.master_addr = master_addr
        self.master_port = master_port
        self.rank = rank
        self.world_size = world_size
        self.backend = backend
        self.device = device
        self.initialized = False
        self.commGrpHandlerDicts = {} # maps from <jobName> to <commGrpHandlerDict>
        return
    
    def init_comm_group_if_not(self):
        if not self.initialized:
            init_method = 'tcp://%s:%d'%(self.master_addr, self.master_port)
            args = {"backend": self.backend, "init_method": init_method, "rank": self.rank, "world_size": self.world_size}
            Logger.log("default comm group initialization started. args: %s" % str(args), flush=True)
            dist.init_process_group(self.backend, init_method=init_method, rank=self.rank, world_size=self.world_size)
            assert dist.get_world_size() == self.world_size
            Logger.log("default comm group initialized.", flush=True)
            self.initialized = True
            Logger.log("Testing c10d backend..", flush=True)
            self.testRing()

    def testRing(self):
        dstRank = (self.rank + 1) % self.world_size
        tag = 0
        tensor2send = torch.tensor(self.rank, dtype=torch.int)
        if self.backend == 'nccl':
            Logger.log("[CommunicationBackend] currently testRing deadlocks with nccl. Skipping test.", level=2, flush=True)
            return
            tensor2send = tensor2send.to(self.device)
        
        argsToPrint = {"tensor": tensor2send, "dst": dstRank, "tag": tag}
        Logger.log("[CommunicationBackend] testRing dist.send(%s)"%str(argsToPrint), level=0, flush=True)
        sendReq = dist.isend(tensor=tensor2send, dst=dstRank, tag=tag)
        
        srcRank = (self.world_size + self.rank - 1) % self.world_size
        tensor2recv = torch.zeros(1, dtype=torch.int)
        if self.backend == 'nccl':
            tensor2recv = tensor2recv.to(self.device)
        
        recvReq = dist.irecv(tensor=tensor2recv, src=srcRank, tag=tag)
        sendReq.wait()
        recvReq.wait()
        Logger.log("[CommunicationBackend] testRing completed. received %s"%str(tensor2recv), level=0, flush=True)
    
    def initCommGroups(self, jobName: str, commGrpDict: dict):
        commGrpHandlerDict = {}
        for grpName in commGrpDict:
            globalGrpRanks = commGrpDict[grpName]
            grpHandler = dist.new_group(globalGrpRanks)
            commGrpHandlerDict[grpName] = grpHandler
        self.commGrpHandlerDicts[jobName] = commGrpHandlerDict
        assert 'all' in self.commGrpHandlerDicts[jobName]
        Logger.log("Testing default collective comm all-reduce for %s among %d ranks." % (jobName, len(commGrpDict['all'])), flush=True)
        self.testAllGroupComm(jobName)

    def testAllGroupComm(self, jobName: str):
        commGrpHandlerDict = self.commGrpHandlerDicts[jobName]
        commGrpHandler = commGrpHandlerDict['all']
        tsr = torch.ones(2, dtype=torch.int)
        if self.backend == 'nccl':
            tsr = tsr.to(self.device)
        Logger.log("[CommunicationBackend] testAllGroupComm started for %s. Exchange %s" % (jobName, str(tsr)), flush=True)
        dist.all_reduce(tsr, dist.ReduceOp.SUM, commGrpHandler)
        Logger.log("[CommunicationBackend] testAllGroupComm completed for %s. Receive %s" % (jobName, str(tsr)), flush=True)

    def makeCommunicationHandler(self, jobName, worldSize, tensor_tags, jobRankToGlobalRank):
        sendFromCpu = (self.backend == 'gloo')
        deviceForComm = 'cpu' if self.backend == 'gloo' else self.device
        Logger.log("[CommunicationBackend] makeCommunicationHandler sendFromCpu(%s)"%str(sendFromCpu), level=0)
        if jobName not in self.commGrpHandlerDicts:
            Logger.log("Error in makeCommunicationHandler. commGroupHandlers are not previously initialized for %s." % jobName, level=2, flush=True)
        commGrpHandlerDict = self.commGrpHandlerDicts.pop(jobName)
        return CommunicationHandler(worldSize, tensor_tags, jobRankToGlobalRank, sendFromCpu, deviceForComm, commGrpHandlerDict, shouldSendSizes=True)

class CommunicationHandler:
    # Features.
    # - mapping from a rank for a training job to global runtime rank.
    # - keeps tensor dimension information for recv operation. (c10d recv needs a tensor with correct size)
    def __init__(self, worldSize, tensor_tags, jobRankToGlobalRank, sendFromCpu, deviceForComm, commGrpHandlerDict, shouldSendSizes: bool = True):
        self.tensorSizes = {}
        self.tensor_tags = tensor_tags
        self.jobRankToGlobalRank = jobRankToGlobalRank #list(range(worldSize))
        self.shouldSendSizes = shouldSendSizes
        self.sendFromCpu = sendFromCpu
        self.deviceForComm = deviceForComm
        self.asyncReqs = []
        # self.commGrpDict = commGrpDict
        self.commGrpHandlerDict = commGrpHandlerDict
        # self.addCommGroups(commGrpDict)

    # def initCommGroups(self, commGrpDict):
    #     for grpName in commGrpDict:
    #         grpRanks = commGrpDict[grpName]
    #         globalGrpRanks = [self.jobRankToGlobalRank[rank] for rank in grpRanks]
    #         grpHandler = dist.new_group(globalGrpRanks)
    #         self.commGrpHandlerDict[grpName] = grpHandler

    def stopSendingSizes(self):
        """ Should be called after 1st iteration for performance """
        self.shouldSendSizes = False
    
    def send(self, tensor: torch.Tensor, tensorName: str, dest: int):
        self.sendAsync(tensor, tensorName, dest)
        self.waitForAll()

    def sendAsync(self, tensor: torch.Tensor, tensorName: str, dest: int):
        # assert tensor.is_cuda
        if self.sendFromCpu:
            tensor = tensor.cpu()

        dstRank = self.jobRankToGlobalRank[dest]
        tag = self.tensor_tags[tensorName]
        if self.shouldSendSizes:
            tensor_shape = torch.tensor(tensor.shape, dtype=torch.int, device=self.deviceForComm)
            tensor_shape_len = torch.tensor(len(tensor.shape), dtype=torch.int, device=self.deviceForComm)
            Logger.log("dist.send(%s)"%str({"tensor": tensor_shape_len.size(), "dst": dstRank, "tag": tag}), level=0, flush=True)
            dist.send(tensor=tensor_shape_len, dst=dstRank, tag=tag)
            Logger.log("dist.send(%s)"%str({"tensor": tensor_shape.size(), "dst": dstRank, "tag": tag}), level=0, flush=True)
            dist.send(tensor=tensor_shape, dst=dstRank, tag=tag)
            Logger.log("dist.isend(%s)"%str({"tensor": tensor.size(), "dst": dstRank, "tag": tag, "bytes": tensor.element_size()*tensor.nelement(), "elems": tensor.nelement(), "elemSize": tensor.element_size()}), level=0, flush=True)
        # Logger.log("dist.isend(%s)"%str({"tensor": tensor.size(), "dst": dstRank, "tag": tag}), level=0, flush=True)
        # dist.send(tensor=tensor, dst=dstRank, tag=tag)
        tensorReq = dist.isend(tensor=tensor, dst=dstRank, tag=tag)
        self.asyncReqs.append(tensorReq)
        # return tensorReq

    def waitForAll(self):
        for req in self.asyncReqs:
            req.wait()
        self.asyncReqs.clear()

    def recv(self, tensorName: str, src: int, dtype=torch.float32) -> torch.Tensor:
        self.waitForAll()
        tensor = self.recvAsync(tensorName, src, dtype)
        self.waitForAll()
        if self.sendFromCpu:
            tensor = tensor.cuda()
        return tensor

    def recvAsync(self, tensorName: str, src: int, dtype=torch.float32) -> torch.Tensor:
        src_rank = self.jobRankToGlobalRank[src]
        tag = self.tensor_tags[tensorName]
        if self.shouldSendSizes:
            tensor_shape_len = torch.zeros(1, dtype=torch.int, device=self.deviceForComm)
            Logger.log("dist.recv(%s)"%str({"tensor": tensor_shape_len.size(), "src": src_rank, "tag": tag}), level=0, flush=True)
            dist.recv(tensor=tensor_shape_len, src=src_rank, tag=tag)
            tensor_shape_len = list(map(lambda x: int(x), tensor_shape_len))
            
            tensor_shape = torch.zeros(tensor_shape_len, dtype=torch.int, device=self.deviceForComm)
            Logger.log("dist.recv(%s)"%str({"tensor": tensor_shape.size(), "src": src_rank, "tag": tag}), level=0, flush=True)
            dist.recv(tensor=tensor_shape, src=src_rank, tag=tag)
            tensor_shape = list(map(lambda x: int(x), tensor_shape))

            self.tensorSizes[tensorName] = tensor_shape
        else:
            tensor_shape = self.tensorSizes[tensorName]
        # Receive tensor.
        tensor = torch.empty(tensor_shape, dtype=dtype, device=self.deviceForComm, requires_grad=True)
        # Logger.log("dist.irecv(%s)"%str({"tensor": tensor.size(), "src": src_rank, "tag": tag, "require_grad": tensor.requires_grad}), level=0, flush=True)
        # dist.recv(tensor=tensor, src=src_rank, tag=tag)
        asyncReq = dist.irecv(tensor=tensor, src=src_rank, tag=tag)
        self.asyncReqs.append(asyncReq)
        # Logger.log("dist.irecv(%s)"%str({"require_grad": tensor.requires_grad}), level=0, flush=True)
        return tensor

    def allGather(self, tensorList, tensor, grpName):
        commGrpHandler = self.commGrpHandlerDict[grpName]
        dist.all_gather(tensorList, tensor, commGrpHandler)

    def allReduce(self, tensor, operation, grpName):
        commGrpHandler = self.commGrpHandlerDict[grpName]
        enum = dist.ReduceOp.SUM # operation argument should specify, now default to SUM
        if self.shouldSendSizes:
            Logger.log("dist.all_reduce(%s)"%str({"tensor": tensor.size(), "enum": enum, "grpName": grpName, "kbytes": (tensor.element_size()*tensor.nelement() / 1024)}), level=0, flush=True)
        dist.all_reduce(tensor, enum, commGrpHandler)

File: csrc/communication.cpp
// Copyright (c) 2021 MIT
//
// Permission to use, copy, modify, and distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

#include "communication.h"

#include <torch/torch.h>

#include <memory>
#include <string>

#include "Cycles.h"
#include "json.hpp"
#include "logger.h"
#include "runtime.h"
#include "utils.h"

using Cycles = RAMCloud::Cycles;
using json = nlohmann::json;

/**
 * Constructs communicationHandler base class.
 *
 * \param worldSize   Number of ranks.
 * \param tensorTags  Mapping from xferName to p2p communication tag.
 * \param rank        Rank of the current node.
 * \param jobRankToGlobalRank   Mapping from job's internal rank to cluster
 * rank.
 */
CommunicationHandler::CommunicationHandler(int worldSize, json tensorTags,
                                           int rank, json jobRankToGlobalRank)
    : worldSize(worldSize),
      tensorTags(tensorTags),
      rank(rank),
      jobRankToGlobalRank(jobRankToGlobalRank) {}

int CommunicationHandler::getTag(const std::string& xferName) {
  return tensorTags[xferName].get<int>();
}

///////////////////////////////////////////////////////////////////////
// CommunicationHandlerGRPC
///////////////////////////////////////////////////////////////////////

CommunicationHandlerGRPC::CommunicationHandlerGRPC(std::string taskName,
                                                   int worldSize,
                                                   json tensorTags, int rank,
                                                   json jobRankToGlobalRank)
    : CommunicationHandler(worldSize, tensorTags, rank, jobRankToGlobalRank),
      taskName(taskName) {
  DP_LOG(DEBUG, "Constructing CommunicationHandlerGRPC for %s",
         taskName.c_str());
  rtctx->commHandlerMap[taskName] = this;
  DP_LOG(DEBUG, "set CommunicationHandlerGRPC to commHandlerMap.");
}

/**
 * Save the received data for p2p communication.
 */
void CommunicationHandlerGRPC::saveData(const std::string& tensorData,
                                        int tag) {
  // DP_LOG(DEBUG, "awaiting grpc lock.");
  std::lock_guard<std::mutex> lock(_mutex);
  receivedData[tag] = tensorData;
  _cv.notify_all();
}

/**
 * Implements p2p send (refer to CommunicationHandler::send)
 */
void CommunicationHandlerGRPC::send(const torch::Tensor& tensor, int tag,
                                    int dest) {
  RuntimeClient* destClient;
  {
    DP_LOG(DEBUG, "Grabbing a lock");
    std::lock_guard<std::mutex> lock(_mutex);
    auto search = clientPool.find(dest);
    if (search == clientPool.end()) {
      std::string ipAndPort = rtctx->rankToIpAndPort[dest];
      DP_LOG(DEBUG, "Dest:%d(%s) isn't in clientPool yet.", dest,
             ipAndPort.c_str());

      auto channel =
          grpc::CreateChannel(ipAndPort, grpc::InsecureChannelCredentials());
      auto client = std::make_unique<RuntimeClient>(channel);
      clientPool[dest] = std::move(client);
      clientPool[dest]->Poke();
      DP_LOG(DEBUG, "Poked dest:%d", dest);
    }
    destClient = clientPool[dest].get();
    DP_LOG(DEBUG, "Releasing a lock");
  }

  std::string tsrData;
  tsrData.resize(tensor.nbytes());
  c10::cuda::device_synchronize(); /* ensure data is ready to be copied out */
  CUDACHECK(cudaMemcpy(&tsrData[0], tensor.data_ptr(), tensor.nbytes(),
                       cudaMemcpyDefault));
  DP_LOG(DEBUG, "Copied tensor data (potentially CUDA) to CPU.");
  destClient->P2PCommunication(taskName, tsrData, tag);
}

/**
 * Implements p2p recv (refer to CommunicationHandler::recv)
 */
void CommunicationHandlerGRPC::recv(torch::Tensor& tensor, int tag, int src) {
  UNUSED(src);
  DP_LOG(DEBUG, "Grabbing a lock");
  std::unique_lock<std::mutex> lock(_mutex);
  _cv.wait(lock, [&] { return receivedData.count(tag); });
  auto search = receivedData.find(tag);
  assert(search != receivedData.end());
  std::string tensorData = search->second;
  assert(tensor.nbytes() == tensorData.size());
  CUDACHECK(cudaMemcpyAsync(tensor.data_ptr(), tensorData.data(),
                            tensorData.size(), cudaMemcpyHostToDevice,
                            rtctx->torch_stream));
  receivedData.erase(search);
  DP_LOG(DEBUG, "Releasing a lock");
}

void CommunicationHandlerGRPC::all_reduce(torch::Tensor& tensor,
                                          c10d::ReduceOp op) {
  UNUSED(tensor);
  UNUSED(op);
  fprintf(stderr, "GRPC all reduce not implemented\n");
}

void CommunicationHandlerGRPC::testRingP2P() {
  torch::Tensor tsr2Send = torch::ones({2}, rtctx->c10dev);
  // DP_LOG(DEBUG, "Created tensor [%s] in dev.", tsrToStr(tsr2Send).c_str());
  int dest = (rtctx->rank + 1) % rtctx->worldSize;
  send(tsr2Send, 1, dest);
  DP_LOG(DEBUG, "Sent tensor [%s] to %d.", tsrToStr(tsr2Send).c_str(), dest);

  torch::Tensor tsr2Recv = torch::zeros({2}, rtctx->c10dev);
  // DP_LOG(DEBUG, "Created tensor [%s] in dev.", tsrToStr(tsr2Recv).c_str());
  int src = (rtctx->worldSize + rtctx->rank - 1) % rtctx->worldSize;
  recv(tsr2Recv, 1, src);
  DP_LOG(DEBUG, "Rcvd tensor [%s] from %d.", tsrToStr(tsr2Recv).c_str(), src);
}

///////////////////////////////////////////////////////////////////////
// CommunicationHandlerNCCL
///////////////////////////////////////////////////////////////////////

CommunicationHandlerNCCL::CommunicationHandlerNCCL(std::string taskName,
                                                   int worldSize,
                                                   json tensorTags, int rank,
                                                   json jobRankToGlobalRank)
    : CommunicationHandler(worldSize, tensorTags, rank, jobRankToGlobalRank),
      default_comm_stream(c10::cuda::getStreamFromPool(true)) {
  DP_LOG(DEBUG, "Constructing CommunicationHandlerNCCL for %s",
         taskName.c_str());
  rtctx->commHandlerMap[taskName] = this;
  DP_LOG(DEBUG, "set CommunicationHandlerNCCL to commHandlerMap.");
}

void CommunicationHandlerNCCL::sync(
    c10::optional<c10::cuda::CUDAStream> stream) {
  if (in_group_call) return;
  sync_event.record(stream ? stream.value() : default_comm_stream);
  sync_event.block(rtctx->torch_stream);
}

void CommunicationHandlerNCCL::send(const torch::Tensor& tensor, int tag,
                                    int dest) {
  UNUSED(tag);

  assert(in_group_call);
  tensor.record_stream(group_call_stream.value());
  torch::cuda::nccl::send(tensor, group_call_commObj, group_call_stream.value(),
                          dest);
}

void CommunicationHandlerNCCL::recv(torch::Tensor& tensor, int tag, int src) {
  UNUSED(tag);
  DP_LOG(DEBUG, "NCCL recv.");

  assert(in_group_call);
  tensor.record_stream(group_call_stream.value());
  torch::cuda::nccl::recv(tensor, group_call_commObj, group_call_stream.value(),
                          src);
}

static ncclDataType_t to_nccl_data_type(c10::ScalarType type) {
  switch (type) {
    case at::kFloat:
      return ncclDataType_t::ncclFloat;
    case at::kHalf:
      return ncclDataType_t::ncclHalf;
    case at::kDouble:
      return ncclDataType_t::ncclDouble;
    case at::kLong:
      return ncclDataType_t::ncclInt64;
    case at::kInt:
      return ncclDataType_t::ncclInt;
    case at::kChar:
      return ncclDataType_t::ncclChar;
    case at::kByte:
      return ncclDataType_t::ncclUint8;
    case at::kBool:
      return ncclDataType_t::ncclUint8;
    default:
      TORCH_CHECK(false, "Unconvertible NCCL type ", type);
  }
}

void CommunicationHandlerNCCL::all_reduce(torch::Tensor& tensor,
                                          c10d::ReduceOp op) {
  // NCCL op mapping from c10d
  const std::map<c10d::ReduceOp, ncclRedOp_t> ncclOp = {
      {c10d::ReduceOp::MIN, ncclMin},
      {c10d::ReduceOp::MAX, ncclMax},
      {c10d::ReduceOp::SUM, ncclSum},
      {c10d::ReduceOp::PRODUCT, ncclProd},
  };

  assert(in_group_call);
  NCCL_API_CALL(
      ncclAllReduce(tensor.data_ptr(), tensor.data_ptr(), tensor.numel(),
                    to_nccl_data_type(tensor.scalar_type()), ncclOp.at(op),
                    reinterpret_cast<ncclComm_t>(group_call_commObj),
                    group_call_stream.value()));
}

void CommunicationHandlerNCCL::testAllReduce() {
  if (rtctx->worldSize == 1) return;

  /* sum */
  torch::Tensor t = torch::full({3, 3}, rtctx->rank + 1, rtctx->c10dev);
  comm_start();
  all_reduce(t, c10d::ReduceOp::SUM);
  comm_end();
  sync();
  int sum = 0;
  for (int i = 1; i < rtctx->worldSize + 1; i++) sum += i;
  torch::Tensor expected = torch::full({3, 3}, sum, rtctx->c10dev);
  assert(at::equal(t, expected));

  /* prod */
  int prod = 1;
  for (int i = 1; i < rtctx->worldSize + 1; i++) prod *= i;
  t = torch::full({3, 3}, rtctx->rank + 1, rtctx->c10dev);
  expected = torch::full({3, 3}, prod, rtctx->c10dev);
  comm_start();
  all_reduce(t, c10d::ReduceOp::PRODUCT);
  comm_end();
  sync();
  assert(at::equal(t, expected));

  /* max */
  t = torch::full({3, 3}, rtctx->rank + 1, rtctx->c10dev);
  expected = torch::full({3, 3}, rtctx->worldSize, rtctx->c10dev);
  comm_start();
  all_reduce(t, c10d::ReduceOp::MAX);
  comm_end();
  sync();
  assert(at::equal(t, expected));

  DP_LOG(NOTICE, "Completed 3 NCCL all reduce tests.");
}

void CommunicationHandlerNCCL::testRingP2P() {
  if (rtctx->worldSize == 1) return;

  int dest = (rtctx->rank + 1) % rtctx->worldSize;
  int src = (rtctx->rank + rtctx->worldSize - 1) % rtctx->worldSize;

  torch::Tensor send_tensor =
      torch::full({3, 3}, rtctx->rank + 1, rtctx->c10dev);
  torch::Tensor recv_tensor = torch::full({3, 3}, src, rtctx->c10dev);
  torch::Tensor expected = torch::full({3, 3}, src + 1, rtctx->c10dev);

  comm_start();

  if (rtctx->rank == 0) {
    send(send_tensor, 1, dest);
    recv(recv_tensor, 1, src);
  } else {
    recv(recv_tensor, 1, src);
    send(send_tensor, 1, dest);
  }

  comm_end();
  sync();

  DP_LOG(DEBUG, "Sent tensor [%s] to %d.", tsrToStr(send_tensor).c_str(), dest);
  DP_LOG(DEBUG, "Received tensor [%s] from %d.", tsrToStr(recv_tensor).c_str(),
         src);
  DP_LOG(DEBUG, "Expected tensor [%s] from %d.", tsrToStr(expected).c_str(),
         src);

  assert(at::equal(recv_tensor, expected));
}

void CommunicationHandlerNCCL::comm_start(
    c10::optional<c10::cuda::CUDAStream> stream, size_t commKey) {
  if (++in_group_call != 1)
    return;

  if (commKey > 0) {
    auto& config = rtctx->nccl_groups.at(commKey);
    group_call_commObj = config.ncclCommObj;
  } else {
    group_call_commObj = rtctx->maingroup.ncclCommObj;
  }

  group_call_stream = stream ? stream : default_comm_stream;
  sync_event.record(rtctx->torch_stream);
  sync_event.block(group_call_stream.value());
  NCCL_API_CALL(ncclGroupStart());
}

void CommunicationHandlerNCCL::comm_end() {
  if (--in_group_call != 0)
    return;
  group_call_stream = {};
  NCCL_API_CALL(ncclGroupEnd());
}

void CommunicationHandlerNCCL::precapture() {
  sync_event.record(rtctx->torch_stream);
  sync_event.block(default_comm_stream);
}

void CommunicationHandlerNCCL::postcapture() {
  sync_event.record(default_comm_stream);
  sync_event.block(rtctx->torch_stream);
}
